src/wasmedge/wasmedge.cc

```diff
@@ -0,0 +1,592 @@
+  bool link(std::string_view debug_name) override;
+  uint64_t getMemorySize() override;
+  std::optional<std::string_view> getMemory(uint64_t pointer, uint64_t size) override;
+  bool setMemory(uint64_t pointer, uint64_t size, const void *data) override;
+  bool getWord(uint64_t pointer, Word *word) override;
+  bool setWord(uint64_t pointer, Word word) override;
+  size_t getWordSize() override { return sizeof(uint32_t); };
+
+#define _REGISTER_HOST_FUNCTION(T)                                                                 \
+  void registerCallback(std::string_view module_name, std::string_view function_name, T,           \
+                        typename ConvertFunctionTypeWordToUint32<T>::type f) override {            \
+    registerHostFunctionImpl(module_name, function_name, f);                                       \
+  };
+  FOR_ALL_WASM_VM_IMPORTS(_REGISTER_HOST_FUNCTION)
+#undef _REGISTER_HOST_FUNCTION
+
+#define _GET_MODULE_FUNCTION(T)                                                                    \
+  void getFunction(std::string_view function_name, T *f) override {                                \
+    getModuleFunctionImpl(function_name, f);                                                       \
+  };
+  FOR_ALL_WASM_VM_EXPORTS(_GET_MODULE_FUNCTION)
+#undef _GET_MODULE_FUNCTION
+private:
+  template <typename... Args>
+  void registerHostFunctionImpl(std::string_view module_name, std::string_view function_name,
+                                void (*function)(Args...));
+
+  template <typename R, typename... Args>
+  void registerHostFunctionImpl(std::string_view module_name, std::string_view function_name,
+                                R (*function)(Args...));
+
+  template <typename... Args>
+  void getModuleFunctionImpl(std::string_view function_name,
+                             std::function<void(ContextBase *, Args...)> *function);
+
+  template <typename R, typename... Args>
+  void getModuleFunctionImpl(std::string_view function_name,
+                             std::function<R(ContextBase *, Args...)> *function);
+
+  WasmEdgeLoaderPtr loader_;
+  WasmEdgeValidatorPtr validator_;
+  WasmEdgeExecutorPtr executor_;
+  WasmEdgeStorePtr store_;
+  WasmEdgeASTModulePtr module_;
+  WasmEdge_MemoryInstanceContext *memory_;
+
+  std::unordered_map<std::string, HostFuncDataPtr> host_functions_;
+  std::unordered_map<std::string, HostModuleDataPtr> import_objects_;
+  std::unordered_set<std::string> module_functions_;
+};
+
+bool WasmEdge::load(std::string_view bytecode, std::string_view,
+                    const std::unordered_map<uint32_t, std::string>) {
+  WasmEdge_ASTModuleContext *mod = nullptr;
+  WasmEdge_Result res = WasmEdge_LoaderParseFromBuffer(
+      loader_.get(), &mod, reinterpret_cast<const uint8_t *>(bytecode.data()), bytecode.size());
+  if (WasmEdge_ResultOK(res)) {
+    res = WasmEdge_ValidatorValidate(validator_.get(), mod);
```

Comment:
Could we failing fast here, i.e.
```
WasmEdge_Result res = WasmEdge_LoaderParseFromBuffer(...)
if (!WasmEdge_ResultOK(res)) {
  return false;
}
```
otherwise, this falls thru all the way until the `return`, which is common source for mistakes.